examples/blocksub/main.go

```diff
@@ -0,0 +1,41 @@
+package main
+
+import (
+	"context"
+	"os"
+
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/flashbots/goutils/blocksub"
+)
+
+var (
+	httpURI  = os.Getenv("ETH_HTTP") // usually port 8545
+	wsURI    = os.Getenv("ETH_WS")   // usually port 8546
+	logJSON  = os.Getenv("LOG_JSON") == "1"
+	logDebug = os.Getenv("DEBUG") == "1"
+)
+
+func main() {
+	logLevel := log.LvlInfo
+	if logDebug {
+		logLevel = log.LvlDebug
+	}
+
+	logFormat := log.TerminalFormat(true)
+	if logJSON {
+		logFormat = log.JSONFormat()
+	}
+
+	log.Root().SetHandler(log.LvlFilterHandler(logLevel, log.StreamHandler(os.Stderr, logFormat)))
+	ch := make(chan *ethtypes.Header)
+	bs := blocksub.NewBlockSub(context.Background(), httpURI, wsURI, ch)
+	err := bs.Start()
+	if err != nil {
+		panic(err)
+	}
+
+	for h := range ch {
```

Comment:
I think noone closes the channel, so it's really difficult to do proper cleanup.
Closing it in block sub on context done would make some sense, but on the other hand it's not really the owner of the object.
i guess making an actual context selecting on it together with the channel would work but it's not the prettiest.

I don't really mean this is an issue in the example, but thinking of how it'll go when used in other code which might want to clean up.

Reply:
Maybe we should do it as go-ethereum client does it: you pass in the channel and receive a subscription

```sub, err := client.SubscribeNewHead(s.ctx, wsHeaderC)```

```
// Subscription represents an event subscription where events are
// delivered on a data channel.
type Subscription interface {
	// Unsubscribe cancels the sending of events to the data channel
	// and closes the error channel.
	Unsubscribe()
	// Err returns the subscription error channel. The error channel receives
	// a value if there is an issue with the subscription (e.g. the network connection
	// delivering the events has been closed). Only one value will ever be sent.
	// The error channel is closed by Unsubscribe.
	Err() <-chan error
}```

Reply:
Hm, I like that